### PR TITLE
core: make ParametrizedAttribute parameters a tuple

### PR DESCRIPTION
**Pull request by superlopuh**
Two problems with the current code:
1. the class is declared as frozen, but a list makes that false, it's mutable in a hidden way
2. the fact that it's frozen means a hash function is generated, but python crashes when it tries to hash a list

### CODE REVIEW

**Review by compor — APPROVED** [review]

**Review by AntonLydike — APPROVED** [review]
Nice work!

**Comment by superlopuh** [comment]
After some discussions with @math-fehr, I rolled back a fair number of changes, to keep the current API working. It means we incur the cost of creating a new tuple with every init, but I'm hoping to change the way this class works in the future, so would make sense to keep the number of API changes to 1.

File: tests/test_attribute_definition.py
Comment (superlopuh) [inline tests/test_attribute_definition.py:1]: some of the changes in this file were necessary due to pyright, so I migrated the whole thing, I hope that's ok

**Review by math-fehr — APPROVED** [review]
Very nice!

**Pull request merged**